client/consensus/grandpa/justification.go

```diff
@@ -19,23 +20,74 @@ var (
 	errBlockNotDescendentOfBase = errors.New("block not descendent of base")
 )
 
-// Justification is a GRANDPA justification for block finality, it includes a commit message and
+// GrandpaJustification is a GRANDPA justification for block finality, it includes a commit message and
 // an ancestry proof including all headers routing all precommit target blocks
 // to the commit target block. Due to the current voting strategy the precommit
 // targets should be the same as the commit target, since honest voters don't
 // vote past authority set change blocks.
 //
 // This is meant to be stored in the db and passed around the network to other
 // nodes, and are used by syncing nodes to prove authority set handoffs.
-type Justification[
+type GrandpaJustification[
 	Hash constraints.Ordered,
 	N constraints.Unsigned,
 	S comparable,
 	ID AuthorityID,
-	H Header[Hash, N]] struct {
+] struct {
 	Round           uint64
 	Commit          finalityGrandpa.Commit[Hash, N, S, ID]
-	VotesAncestries []H
+	VotesAncestries []Header[Hash, N]
+}
+
+// Type used for decoding grandpa justifications (can pass in generic Header type)
+type decodeGrandpaJustification[
+	Hash constraints.Ordered,
+	N constraints.Unsigned,
+	S comparable,
+	ID AuthorityID,
+	H Header[Hash, N],
```

Comment:
Do you need this `H` parameter? I think you are not using it.

Reply:
Yes it is needed to properly decode a grandpa justification. This is a helper struct created to decode a `GrandpaJustification` because the `GrandpaJustification.VoterAncestries` is defined as an interface, which you cannot decode into. By passing in H as a generic constraint here we can define `VoterSncestries` to point to a concrete type to be decoded into